radis/lbl/gpu.py

```diff
@@ -14,605 +16,731 @@ class initData(ctypes.Structure):
-    # cdef float Qv2 = 1 / np.power(1 - np.exp(-c2 * w2 / Tv12), d2)
-    # cdef float Qv3 = 1 / np.power(1 - np.exp(-c2 * w3 / Tv3 ), d3)
-    # cdef float Q = Qr * Qv1 * Qv2 * Qv3
-
-    # iter_params_h.Q = Q
-
-
-def set_constant_memory(var_str):
-    var_h = globals()[var_str + "_h"]
-    memptr_d = cuda_module.get_global(var_str + "_d")
-    ptr = ctypes.cast(ctypes.pointer(var_h), ctypes.c_void_p)
-    struct_size = ctypes.sizeof(var_h)
-    memptr_d.copy_from_host(ptr, struct_size)
+def calc_lorentzian_minmax(param_data, log_rT, log_2p):
+    """
+
+
+    Parameters
+    ----------
+    param_data : TYPE
+        DESCRIPTION.
+    log_rT : TYPE
+        DESCRIPTION.
+    log_2p : TYPE
+        DESCRIPTION.
+
+    Returns
+    -------
+    TYPE
+        DESCRIPTION.
+
+    """
+
+    result = []
+    for params in param_data:
+        A, B, X = params
+        i = 0
+        while X[i] < log_rT:
+            i += 1
+        result.append(log_rT * A[i] + B[i] + log_2p)
+    return tuple(result)
+
+
+def calc_lorentzian_params(param_data, init_h, iter_h, epsilon=1e-4):
+    """
+
+
+    Parameters
+    ----------
+    param_data : TYPE
+        DESCRIPTION.
+    init_h : TYPE
+        DESCRIPTION.
+    iter_h : TYPE
+        DESCRIPTION.
+    epsilon : TYPE, optional
+        DESCRIPTION. The default is 1e-4.
+
+    Returns
+    -------
+    None.
+
+    """
+
+    log_wL_min, log_wL_max = calc_lorentzian_minmax(
+        param_data, iter_h.log_rT, iter_h.log_2p
+    )
+    log_wL_max += epsilon
+    dxL = (log_wL_max - log_wL_min) / (init_h.N_L - 1)
+
+    iter_h.log_wL_min = log_wL_min
+    iter_h.dxL = dxL
+
+
+def set_pTQ(p, T, mole_fraction, Q_arr, iter_h, l=1.0, slit_FWHM=0.0):
+    """
+
+
+    Parameters
+    ----------
+    p : float
+        pressure [bar].
+    T : float
+        temperature [K].
+    mole_fraction : float
+    iter_h : TYPE
+        DESCRIPTION.
+    l : TYPE, optional
+        DESCRIPTION. The default is 1.0.
+    slit_FWHM : TYPE, optional
+        DESCRIPTION. The default is 0.0.
+
+    Returns
+    -------
+    None.
+
+    """
+
+    iter_h.p = p  # bar
+    iter_h.log_2p = np.log(2 * p)
+    iter_h.hlog_T = 0.5 * np.log(T)
+    iter_h.log_rT = np.log(296.0 / T)
+    iter_h.c2T = -c2 / T
+    iter_h.N = mole_fraction * p * 1e5 / (1e6 * k * T)  # cm-3
+    iter_h.l = l
+    iter_h.slit_FWHM = slit_FWHM
+
+    for i in range(len(Q_arr)):
+        iter_h.Q[i] = Q_arr[i]
+
+
+def constant_memory_setter(cuda_module, var_str):
+    def setter(var_h):
+        memptr_d = cuda_module.get_global(var_str)
+        ptr = ctypes.cast(ctypes.pointer(var_h), ctypes.c_void_p)
+        struct_size = ctypes.sizeof(var_h)
+        memptr_d.copy_from_host(ptr, struct_size)
+
+    return setter
 
 
 def gpu_init(
-    v_arr, N_wG, N_wL, iso, v0, da, log_2gs, na, log_2vMm, S0, El, Q, verbose_gpu
+    v_arr,
+    N_G,
+    N_L,
+    iso,
+    v0,
+    da,
+    gamma,
+    na,
+    S0,
+    El,
+    Mm_arr,
```

Comment:
@dcmvdbekerom can you easily pass `dict` as arguments with Cython ? (I hope some, all Python is based on dictionnaries). 
It may be more interesting to use 
```
iso.map(Mm_dictionary)
```
with `Mm_dictionary = {iso:M} ` than 
```
Mm_dictionary.take(iso)
```

- No need to ensure the index are the same  (in your case, if user asks for iso='3' I assume you still have to generate molar_mass & abundance arrays with [1, 2, 3].  Doesn't cost much to transport these arrays though. For partition functions I don't know. 
- it may work with non-integer names of isotopes : HITRAN started using 'A' and 'B' for CO2(11, 12).   No big deal though, we could pre-process this. 

As you see it's very minor. 
@gagan-aryan I wonder if you could try to benchmark the two approaches in terms of performance (I expect very similar results) ? For instance with [perfplot](https://github.com/nschloe/perfplot).

Reply:
I will try it out.

Reply:
Dictionaries in cython are no problem, but it has to be arrays when passing to the GPU.
But this conversion can be done inside gpu_init() so that you can pass dicts as argument to gpu_init().
As I noted in @gagan-aryan PR, we should not just do this for isotopes but also for species while we're at it.

Reply:
Ok good ! 

I'd keep it single-species for the moment ; lineshape parameters may vary a lot and a shared LDM may not be the best; we would better need a LDM "per block" which will require some changes anyway